src/Microsoft.Extensions.DotnetToolDispatcher.Sources/DotnetToolDispatcher.cs

```diff
@@ -0,0 +1,98 @@
+        private const string DispatcherVersionArgumentName = "--dispatcher-version";
+
+        private static readonly string ToolName = PlatformServices.Default.Application.ApplicationName;
+        private static readonly string DispatcherVersionArgumentValue = ResolveDispatcherVersionArgumentValue();
+
+        public static ICommand CreateDispatchCommand(
+            IEnumerable<string> dispatchArgs,
+            ProjectContext projectContext,
+            string configuration,
+            string outputPath,
+            string buildBasePath)
+        {
+            configuration = configuration ?? Constants.DefaultConfiguration;
+            var commandFactory = new ProjectDependenciesCommandFactory(
+                projectContext.TargetFramework,
+                configuration,
+                outputPath,
+                buildBasePath,
+                projectContext.ProjectDirectory);
+
+            var dispatchArgsList = new List<string>(dispatchArgs);
+            dispatchArgsList.Add(DispatcherVersionArgumentName);
+            dispatchArgsList.Add(DispatcherVersionArgumentValue);
+
+            var command = commandFactory.Create(
+                ToolName,
+                dispatchArgsList,
+                projectContext.TargetFramework,
+                configuration);
+            return command;
+        }
+
+        public static bool IsDispatcher(string[] programArgs) =>
+            !programArgs.Contains(DispatcherVersionArgumentName, StringComparer.OrdinalIgnoreCase);
+
+        public static void EnsureValidDispatchRecipient(CommandLineApplication application, string[] programArgs)
+        {
+            Debug.Assert(!IsDispatcher(programArgs));
+
+            if (!application.Options.Any(option =>
+                string.Equals(option.Template, DispatcherVersionArgumentName, StringComparison.OrdinalIgnoreCase)))
+            {
+                // Add the dispatch argument to ensure no errors are caused by unhandled arguments.
+                application.Option(
+                    DispatcherVersionArgumentName,
+                    "A flag indicating if the tool should dispatch its execution.",
+                    CommandOptionType.SingleValue);
+            }
+
+            var dispatcherArgumentIndex = Array.FindIndex(
+                programArgs,
+                (value) => string.Equals(value, DispatcherVersionArgumentName, StringComparison.OrdinalIgnoreCase));
+            var dispatcherArgumentValueIndex = dispatcherArgumentIndex + 1;
+            if (dispatcherArgumentValueIndex < programArgs.Length)
+            {
+                var dispatcherVersion = programArgs[dispatcherArgumentValueIndex];
+
+                if (string.Equals(dispatcherVersion, DispatcherVersionArgumentValue, StringComparison.Ordinal))
+                {
+                    return;
+                }
+            }
+
+            // Could not validate the dispatchers version.
+            throw new InvalidOperationException(
```

Comment:
Exactly..... Still hypothetical, but something like https://github.com/natemcmaster/dotnet-command could in theory still work with dotnet-razor-tooling :wink: